# Clustering Project - German U-Boats of World War II

In [3]:
import numpy as np
import pandas as pd
import sklearn.preprocessing, sklearn.cluster, sklearn.metrics
import scipy.spatial
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
import plotly.express as px
from scipy.spatial import distance_matrix
from scipy.cluster.hierarchy import dendrogram
from sklearn.cluster import AgglomerativeClustering, DBSCAN
import matplotlib.pyplot as plt
import seaborn as sns
import requests, zipfile, io
from sklearn.cluster import DBSCAN
from sklearn.cluster import AffinityPropagation
from sklearn.cluster import OPTICS
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import silhouette_score, davies_bouldin_score, calinski_harabasz_score, make_scorer
from plotly.subplots import make_subplots
from sklearn.mixture import GaussianMixture
from sklearn.model_selection import GridSearchCV
from scipy.stats import pearsonr, spearmanr

##### 
- Zdroj dat: https://www.kaggle.com/datasets/cormac42/ww2-u-boats
- Dataset obsahuje informace o německých ponorkách z druhé světové války.
- Data jsou scrapována z wikipedia stránek.


## Preprocessing

In [5]:
# Load the data
df = pd.read_csv('uboats.csv')
df

Name  Year   Type                                 Notable Commanders  \
0        U-1  1935    IIA                                       Klaus Ewerth   
1        U-2  1935    IIA  Hans Heidtmann,Heinrich Liebe,Helmut Rosenbaum...   
2        U-3  1935    IIA  Joachim Schepke,Otto von Bülow,Hans-Hartwig Tr...   
3        U-4  1935    IIA                                Heinz-Otto Schultze   
4        U-5  1935    IIA                       Heinrich Lehmann-Willenbrock   
...      ...   ...    ...                                                ...   
1148  U-4707  1945  XXIII                                                NaN   
1149  U-4709  1945  XXIII                                                NaN   
1150  U-4710  1945  XXIII                              Ludwig von Friedeburg   
1151  U-4711  1945  XXIII                                                NaN   
1152  U-4712  1945  XXIII                                                NaN   

      Warships_sunk_n_total_loss_No Warships_sunk_n_total_loss_Tons-n-GRT  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 1                                 1,090   
4                                 0                                     0   
...                             ...                                   ...   
1148                              0                                     0   
1149                              0                                     0   
1150                              0                                     0   
1151                              0                                     0   
1152                              0                                     0   

      Warships_Damaged_No Warships_Damaged_Tons-n-GRT  Merchant_Ships_sunk_No  \
0                       0                           0                       0   
1                       0                           0                       0   
2                       0                           0                       2   
3                       0                           0                       3   
4                       0                           0                       0   
...                   ...                         ...                     ...   
1148                    0                           0                       0   
1149                    0                           0                       0   
1150                    0                           0                       0   
1151                    0                           0                       0   
1152                    0                           0                       0   

     Merchant_Ships_sunk_GRT  ...          Notes  \
0                          0  ...  Struck a mine   
1                          0  ...  Training boat   
2                      2,348  ...            NaN   
3                      5,133  ...            NaN   
4                          0  ...       Accident   
...                      ...  ...            ...   
1148                       0  ...            NaN   
1149                       0  ...            NaN   
1150                       0  ...            NaN   
1151                       0  ...            NaN   
1152                       0  ...            NaN   

                                                    URL  Commissioned  \
0     https://en.wikipedia.org/wiki/German_submarine...    1935-06-29   
1     https://en.wikipedia.org/wiki/German_submarine...    1935-07-25   
2     https://en.wikipedia.org/wiki/German_submarine...    1935-09-06   
3     https://en.wikipedia.org/wiki/German_submarine...    1935-08-17   
4     https://en.wikipedia.org/wiki/German_submarine...    1935-08-31   
...                                                 ...           ...   
1148  https:/

In [15]:
# print table where will be each column with its name, type, number of missing values and unique values
def print_columns_descrbition(df):
    tab = pd.DataFrame(columns=['column_name', 'type', 'missing', 'unique'])
    for col in df.columns:
        new_row = pd.DataFrame({'column_name': [col], 'type': [df[col].dtype], 'missing': [df[col].isnull().sum()], 'unique': [df[col].nunique()]})
        tab = pd.concat([tab, new_row], ignore_index=True)
    return tab

# print number of rows and columns
print(f"Table with some description of columns:\nOriginal dataset has {df.shape[0]} rows and {df.shape[1]} columns")
print_columns_descrbition(df)

Table with some description of columns:
Original dataset has 1153 rows and 26 columns


column_name    type missing unique
0                                    Name  object       0   1153
1                                    Year   int64       0     11
2                                    Type  object       0     24
3                      Notable Commanders  object     935    169
4           Warships_sunk_n_total_loss_No   int64       0      7
5   Warships_sunk_n_total_loss_Tons-n-GRT  object       0    156
6                     Warships_Damaged_No   int64       0      4
7             Warships_Damaged_Tons-n-GRT  object       0     46
8                  Merchant_Ships_sunk_No   int64       0     39
9                 Merchant_Ships_sunk_GRT  object       0    423
10              Merchant_Ships_damaged_No   int64       0      7
11             Merchant_Ships_damaged_GRT  object       0    188
12           Merchant_Ships_total_loss_No   int64       0      4
13          Merchant_Ships_total_loss_GRT  object       0     49
14                             Fate_Event  object       0     17
15                              Fate_Date  object       3    548
16                                  Notes  object     250    482
17                                    URL  object       0   1153
18                           Commissioned  object       2    803
19                                Patrols  object     296    852
20                          Patrols_Count   int64       0     20
21                              Wolfpacks  object     779    364
22                        Wolfpacks_Count   int64       0     21
23                               Flotilla  object       0   1147
24                         Flotilla_Count   int64       0      7
25                          Last_Flotilla  object       0     34

In [16]:
# print fate events unique values and their counts
df['Fate_Event'].value_counts()

Fate_Event
Sunk                       650
Scuttled                   268
Surrendered                149
Missing                     45
Stricken                    15
Captured                     6
Scrapped                     5
Broken up                    4
Buried                       3
Decommissioned               1
Sold                         1
Ran aground & destroyed      1
Interned, then sold          1
Destroyed                    1
Interned                     1
Grounded                     1
Aground                      1
Name: count, dtype: int64

### Popis datasetu
- Dataset obsahuje 1153 ponerek vyrobených Německem od roku 1935 do roku 1945.
- Neobsahuje ponorky, které byly německem ukořistěny a použity německým námořnictvem.

### Popis sloupců
- **Name** - název ponorky
- **Year** - rok, kdy byla ponorka spuštěna na vodu
- **Type** - typ ponorky
- **Notable Commanders** - významní velitelé ponorky
- **Warships_sunk_n_total_loss_No** - počet potopených válečných lodí
- **Warships_sunk_n_total_loss_Tons-n-GRT** - celková tonáž potopených válečných lodí
- **Warships_Damaged_No** - počet válečných lodí, které byly poškozeny
- **Warships_Damaged_Tons-n-GRT** - celková tonáž válečných lodí, které byly poškozeny
- **Merchant_Ships_sunk_No** - počet potopených obchodních lodí
- **Merchant_Ships_sunk_GRT** - celková tonáž potopených obchodních lodí
- **Merchant_Ships_Damaged_No** - počet obchodních lodí, které byly poškozeny
- **Merchant_Ships_Damaged_GRT** - celková tonáž obchodních lodí, které byly poškozeny
- **Merchant_Ships_total_loss_No** - počet obchodních lodí, které byly ztraceny
- **Merchant_Ships_total_loss_GRT** - celková tonáž obchodních lodí, které byly ztraceny
- **Fate_Event** - osud ponorky
- **Fate_Date** - datum, kdy k osudu ponorky došlo
- **Notes** - poznámky(Např. název lodě, která ponorku potopila; způsob potopení; způsob použití; atd.)

In [ ]:
# TODO - drop columns which will not be used
# TODO - missing values
# TODO - encoding
# TODO - scaling
# TODO - slocit sunk and destroyeD?
# TODO - correlation
# TODO - outliers
# TODO - popis co jsme zatim zjistili

In [ ]:
# TODO - clustering 3 methods (kmeans...)
# - on scaled and unscaled data
# only with numerical columns 